### DISCUSSION TITLE
Sistema múltiplos modulos

### DISCUSSION THREAD

**Discussion by jdgabriel** [frontendbr/forum#2347] [Perguntas e Respostas] [answered]
Fala galera, tudo bão?
Fui contrato por uma empresa de Engenharia Elétrica, que possui vários sistemas internos (RH, Projetos, Estoque e etc).
Recebi a missão de refatorar e melhorar o backend e o frontend. Hoje são 5 módulos(webs/pastas) dentro de um projeto em Angular, e um backend para todos os módulos.

O que vcs fariam?

😊

**Comment by felquis** [comment, marked as answer]
Olha, talvez essa seja uma resposta muito genérica, porém quero compartilhar com você.
refatorar não é algo que a gente faz só porque não tem outra coisa melhor para fazer.
Talvez a melhor coisa a se fazer, é não fazer nada. E está tudo bem.

---

Geralmente, meu trabalho como front-end é voltando a entender se as pessoas confiam e estão felizes em usar o programa, e vou 
melhorando o programa para que haja um aumento de satisfação das pessoas internas usando essas ferramentas, exemplos: um clique que tem que dar duas vezes; um CPF que tem que remover os pontos manualmente; um relatório que não tem uma coluna importante; é o tempo de carregamento que é lento, um clique que demora e faz as pessoas virarem o olho etc; um erro constante e conhecido, entre outros.

Se você encontrar estes pontos de melhoria, e conseguir priorizar com o time e líderes de projeto, isso quer dizer que você vai estar trabalhando em algo que tem impacto direto na experiência do dia-a-dia das pessoas usando estes projetos internos, e isso trará uma ótima percepção das pessoas com relação a equipe que vc trabalha.

Essas melhorias irão te expor a muitas partes distintas desses projetos, e essa exposição toda dependendo do projeto, e da sua experiência com as tecnologias usadas, em 3 a 6 meses ou 1 ano vão te mostrar muitos pontos passíveis de refactor que podem tornar sua experiência de desenvolvimento muito melhor, mais previsível, e segura, o que vai melhorar o projeto como um todo. E que também podem melhorar a velocidade percebida da aplicação la na ponta onde as pessoas estão usando-o.

---

Então, primeiro entender como estes projetos ajudam as pessoas a trabalharem melhor. Depois, foco em remover duplicidades, melhorar a previsibilidade do código, corrigir erros conhecidos, nome mais descritivos, melhoria de performance/velocidade de diferentes partes, introdução de testes ou incremento de asserts importantes, procurar ferramentas que analisam o código e te alertam de melhorias como eslint, pode ser que exista algo especifico ao angular.

---

Enfim, acredito que haja uma resposta puramente técnica, em forma de lista a sua pergunta, porém para mim apenas se expondo ao código, ao projeto, e percebendo como ele foi feito, é que se pode chegar a essa lista de possíveis melhorias. E nesse ponto, de como se fazer algo melhor com Angular, eu particularmente buscaria as resposta nas documentação do angular, que além de oferecer o framework de JavaScript no front, também tem diversos guias importantes de estruturas de pastas e arquivos, diferentes de outros frameworks que apenas oferecem a lib e vc se vira com o resto.

---

PS: sobre o backend, também vou por uma linha parecida, entender se as pessoas confiam, como atendem as expectativas delas, ver se cai muito, quais são os problemas conhecidos, e tentar focar em algo que se solucionado, possa melhorar o cotidiano dessas pessoas que precisam desse backend de pé, servindo todo mundo. E pode ser que para entender como este backend é amarrado tudo junto, pode ser que você precise fazer algumas tarefas menores que te exponha a partes distintas deste projeto, para que assim você adquira uma visão de como o projeto é estruturado e consiga propor sugestões de melhorias passíveis de refactor.

---

Caso você veja algo no projeto e consiga fazer uma pergunta específica usando algum trecho de código, estrutura de pastas, sem expor nada privado da empresa é claro - coloca aqui no fórum pois será bem interessante acompanhar este processo.

Espero que te inspire, e que outras pessoas mais focadas no angular possam aparecer aqui. Um abraço e boa sorte.

**Reply by jdgabriel** [reply]
Primeiramente, muito obrigado por essa resposta. (Muito boa mesmo 👍)
Ainda estou no meu desenvolvimento como programador, apesar de já ter alguma experiência, e estou estudando muito sobre "como fazer códigos mais claros e limpos". 

O código que me deram não me passa confiança e é difícil manutenção, não segue nenhum padrão. Por isso penso em refatorar o que é possível e tentar de alguma forma melhorar. Atualmente não temos testes, funções com 200+ linhas que fazem de tudo um pouco, pouco código é reaproveitado.

Toda vez que abro o código penso em começar do zero, minha vontade de aplicar o que estou aprendendo sempre me vem a cabeça.

Mas, obrigado novamente pela resposta.

**Reply by felquis** [reply]
De nada :D eu vou falar mais.

Muito bem, é importante você também fazer este levantamento dos por quês das coisas:

- **O código que me deram não me passa confiança**. por quê? Faz uma listinha.
- **E é difícil manutenção**. por quê? O que exatamente torna difícil fazer algo no projeto.
- **Não segue nenhum padrão**. Talvez os guias de padrão do próprio angular dará uma boa direção para você - https://angular.io/guide/styleguide
- **Atualmente não temos testes**. geralmente projetos já "prontos", porém sem testes, podem se beneficiar mais de testes E2E do que Unitários, escrever o e2e antes de começar um refactor, pode ajudar a não quebrar nada importante.
- **Funções com 200+ linhas que fazem de tudo um pouco*. uma função que tem 200 linhas não necessariamente é um problema, mas certamente se você quiser diminuir você conseguiria. Mas existe um "conceito" de que arquivos grandes, ou funções grandes são um problemas o que pode ser, ou não, realmente um problema kkk famoso depende.
- **Pouco código é reaproveitado**. Se você conseguir fazer uma cirurgia, deduplicando um trecho de código, garanta que este seu novo código, vá de encontro com um styleguide claro.

Essas listas vão te ajudar no médio prazo, entender se suas ações estão ajudando o projeto a melhorar de fato, não só para você mas para o time :D 

Também outra dica, tem que ver se o projeto realmente não tem nenhum padrão, **pois as vezes a ordem se encontra em meio ao caos** e você quebrar com essa ordem desorganizada pode tornar o projeto ainda pior do que já está.

Se você é novo no projeto, e novo também na área como um todo. Existe uma dica geral que se aplica, que é tentar entender as circunstâncias que aquele código foi escrito, e entender se existe alguma tentativa de criar um padrão mas que falhou e desandou no meio do caminho, pode ser que você se torne um gatekeeper daquele padrão que foi tentado no inicio e passe a nutri-lo.

Mais uma prosa ai, que é algo que tento aplicar nas coisas que eu toco, espero que te inspire.

**Reply by jdgabriel** [reply]
O que tenho são projetos plenamente MVC, em o V (...rsrs) no backend e no frontend, apenas arquivos organizados em pastas (Sem boa práticas ou padrões). Tenho dificuldades de ver o que realmente tenho que fazer, pelo fato de todos os conceitos e estruturas fugirem dos padrões estabelecidos, dos cursos e posts/blogs, como boas práticas.

Vou ficar lendo e relendo suas respostas até achar um ponto em comum com a ideia que me deram. 
Obrigado.
